File: Sources/AINotebookCore/Cosine.swift
import Accelerate
import Foundation

public enum Cosine {

    /// Cosine similarity in [-1, 1]. Returns 0 when either input is zero-magnitude
    /// or the dimensions don't match.
    public static func similarity(_ a: [Float], _ b: [Float]) -> Float {
        guard a.count == b.count, !a.isEmpty else { return 0 }
        var dot: Float = 0
        vDSP_dotpr(a, 1, b, 1, &dot, vDSP_Length(a.count))
        var magA: Float = 0
        var magB: Float = 0
        vDSP_svesq(a, 1, &magA, vDSP_Length(a.count))
        vDSP_svesq(b, 1, &magB, vDSP_Length(b.count))
        let denom = sqrtf(magA) * sqrtf(magB)
        guard denom > 0 else { return 0 }
        return dot / denom
    }
}

File: windows/src/AINotebook.Core/Rag/Cosine.cs
namespace AINotebook.Core.Rag;

/// <summary>
/// Cosine similarity in [-1, 1]. Returns 0 when either input is zero-magnitude
/// or the dimensions don't match. 1:1 port of Sources/AINotebookCore/Cosine.swift.
/// </summary>
public static class Cosine
{
    public static float Similarity(float[] a, float[] b)
    {
        if (a.Length != b.Length || a.Length == 0)
        {
            return 0f;
        }

        float dot = 0f;
        float magA = 0f;
        float magB = 0f;
        for (int i = 0; i < a.Length; i++)
        {
            dot += a[i] * b[i];
            magA += a[i] * a[i];
            magB += b[i] * b[i];
        }

        float denom = MathF.Sqrt(magA) * MathF.Sqrt(magB);
        if (denom <= 0f)
        {
            return 0f;
        }

        return dot / denom;
    }
}
